packages/next/src/client/components/router-reducer/prefetch-cache-utils.ts

```diff
@@ -9,28 +8,112 @@ import {
+  // search params so that multiple URLs with the same search parameter can re-use
+  // loading states.
+  let pathnameFromUrl = url.pathname
+
+  // RSC responses can differ based on search params, specifically in the case where we aren't
+  // returning a partial response (ie with `PrefetchKind.AUTO`).
+  // In the auto case, since loading.js & layout.js won't have access to search params,
+  // we can safely re-use that cache entry. But for full prefetches, we should not
+  // re-use the cache entry as the response may differ.
+  if (includeSearchParams) {
+    // if we have a full prefetch, we can include the search param in the key,
+    // as we'll be getting back a full response. The server might have read the search
+    // params when generating the full response.
+    pathnameFromUrl += url.search
+  }
 
-  // nextUrl is used as a cache key delimiter since entries can vary based on the Next-URL header
-  if (nextUrl) {
-    return `${nextUrl}%${pathnameFromUrl}`
+  if (prefix) {
+    return `${prefix}${INTERCEPTION_CACHE_KEY_MARKER}${pathnameFromUrl}`
   }
 
   return pathnameFromUrl
 }
 
+function createPrefetchCacheKey(
+  url: URL,
+  kind: PrefetchKind | undefined,
+  nextUrl?: string | null
+) {
+  return createPrefetchCacheKeyImpl(url, kind === PrefetchKind.FULL, nextUrl)
+}
+
+function getExistingCacheEntry(
+  url: URL,
+  kind: PrefetchKind = PrefetchKind.TEMPORARY,
+  nextUrl: string | null,
+  prefetchCache: Map<string, PrefetchCacheEntry>
+): AliasedPrefetchCacheEntry | undefined {
+  // We first check if there's a more specific interception route prefetch entry
+  // This is because when we detect a prefetch that corresponds with an interception route, we prefix it with nextUrl (see `createPrefetchCacheKey`)
+  // to avoid conflicts with other pages that may have the same URL but render different things depending on the `Next-URL` header.
+  for (const maybeNextUrl of [nextUrl, null]) {
+    const cacheKeyWithParams = createPrefetchCacheKeyImpl(
+      url,
+      true,
+      maybeNextUrl
+    )
+    const cacheKeyWithoutParams = createPrefetchCacheKeyImpl(
+      url,
+      false,
+      maybeNextUrl
+    )
+
+    // We check for the cache entry with search params first, as it's more specific.
+    const cacheKeyToUse = url.search
+      ? cacheKeyWithParams
+      : cacheKeyWithoutParams
+
+    if (prefetchCache.has(cacheKeyToUse)) {
+      return prefetchCache.get(cacheKeyToUse)
+    }
+  }
+
+  // If we've gotten to this point, we didn't find a specific cache entry that matched
+  // the request URL.
+  // We attempt a partial match by checking if there's a cache entry with the same pathname.
+  // Regardless of what we find, since it doesn't correspond with the requested URL, we'll mark it "aliased".
+  // This will signal to the router that it should only apply the loading state on the prefetched data.
+  if (kind !== PrefetchKind.FULL) {
+    for (const cacheEntry of prefetchCache.values()) {
```

Comment:
hmm. this works, but i'm wondering if maybe we should make a fast path for the `want: /abc?q=1, have: /abc` case? that's just one map lookup that we could easily do before looping over everything (`prefetchCache.get(cacheKeyWithoutParams)`), and with Form it's going to be common